src/ast/proof.rs

```diff
@@ -0,0 +1,556 @@
+    Require(Condition<'a>),
+
+    /// A contract indicates a promise by the function; if a certain condition is met of the
+    /// inputs, a separate condition will be met by the output.
+    ///
+    /// Conditions in contracts are separated by the 'implies' operator, `=>`. Here is an example:
+    /// ```text
+    /// # x < 4 => _ < 3
+    /// ```
+    /// Note that the return value is referred to by the identifier `_`.
+    Contract {
+        /// The pre-condition
+        ///
+        /// This is optional in order to allow unconditional return values. This could be written
+        /// as the following:
+        /// ```text
+        /// # => _ >= x
+        /// ```
+        /// where the return value is given to be always greater than or equal to some input `x`.
+        pre: Option<Condition<'a>>,
+
+        /// The post-condition
+        post: Condition<'a>,
+    },
+
+    /// A sentinel value for signaling a failure to parse a statement, when we don't want to give a
+    /// hard error.
+    Malformed,
+}
+
+/// A generic condition, e.g. `x + 4 < y` or `3 * _ >= z`.
+///
+/// A condition is broken into three parts: two expressions and a comparison operator between them.
+/// The following diagram illustrates each of the parts of the condition for the example of
+/// `x + 4 < y`:
+/// ```text
+///   x + 4 < y
+///   ──┬── ┬ ┬
+///    lhs  │ └ rhs
+///        op
+/// ```
+///
+/// It is worth noting that many *syntactically valid* conditions will never be semantically valid.
+/// For example: `x - 4` is a syntactically valid condition but cannot be semantically valid
+/// because there is no actual condition on `x`.
+#[derive(Debug, Clone)]
+pub struct Condition<'a> {
+    pub kind: ConditionKind<'a>,
+    source: &'a [Token<'a>],
+}
+
+#[derive(Debug, Clone)]
+pub enum ConditionKind<'a> {
+    /// A condition composed of two conditions, joined by a logical operator
+    Compound {
+        lhs: Box<Condition<'a>>,
+        op: LogicOp,
+        rhs: Box<Condition<'a>>,
+    },
+    /// A simple condition; two expressions, joined by a comparison operator
+    Simple {
+        lhs: Expr<'a>,
+        op: CompareOp,
+        rhs: Expr<'a>,
+    },
+    /// A sentinel value for a failure to parse
+    Malformed,
+}
+
+/// A simple arithmetic expression
+///
+/// These expressions are always composed either of a single value (a constant or a named variable)
+/// or a compound expression, composed of an arithmetic operator and two sub-expressions.
+#[derive(Debug, Clone)]
+pub struct Expr<'a> {
+    pub kind: ExprKind<'a>,
+    source: &'a [Token<'a>],
+}
+
+#[derive(Debug, Clone)]
+pub enum ExprKind<'a> {
+    Compound {
+        lhs: Box<Expr<'a>>,
+        op: ArithOp,
+        rhs: Box<Expr<'a>>,
+    },
+
+    Named(Ident<'a>),
+
+    // TODO: It might be the case that this shouldn't be an isize, so that we can allow values
+    // above isize::MAX in proof statments, but this should be discussed further.
+    Literal(isize),
```

Comment:
This does need discussion. `isize` is definitely fine for now though.

Reply:
That sounds like a reasonable plan